### PR TITLE
JDK-8272570: C2: crash in PhaseCFG::global_code_motion

### PR DESCRIPTION
**Pull request by casparcwang**
In PhaseCFG::global_code_motion, it allocate recalc_pressure_nodes array. condition is at least one block has more than 10 nodes:
```
  if (OptoRegScheduling) {
    for (uint i = 0; i < number_of_blocks(); i++) {
      Block* block = get_block(i);
      if (block->number_of_nodes() > 10) {
        block_size_threshold_ok = true;
        break;
      }
    }
  }
```

In PhaseCFG::select, it uses recalc_pressure_nodes array, if block's node exceed 10:
```
bool block_size_threshold_ok = (block->number_of_nodes() > 10) ? true : false;
```

In this case, block#4's node size is 10 at GCM begin, but increased to 11 later. PhaseCFG::select use null recalc_pressure_nodes array, this leads to crash.
```
# --- schedule_local B4, before: ---
#   31: Region              31    6
#   22: MergeMem          NULL    0   17    0   23   25 NULL NULL
#   25: storeImmB            3   17   24 NULL
#   23: storeImmB            3   17   24 NULL
#    3: MachProj             4
#    4: CallLeafDirect      31    0   17   18    0   19    0 NULL
#    5: MachProj             4
#   19: loadConD          NULL   20 NULL NULL
#   20: MachConstantBaseNode   11 NULL NULL NULL
#    2: Ret                  3   21   22   18   26 NULL NULL NULL
#
#   ready cnt: -2  latency:  0    31: Region
#   ready cnt:  2  latency:  0    22: MergeMem
#   ready cnt:  1  latency:  0    25: storeImmB
#   ready cnt:  1  latency:  0    23: storeImmB
#   ready cnt:  1  latency:  0     3: MachProj
#   ready cnt:  1  latency:  0     4: CallLeafDirect
#   ready cnt:  1  latency:  0     5: MachProj
#   ready cnt:  1  latency:  0    19: loadConD
#   ready cnt:  0  latency:  0    20: MachConstantBaseNode
#   ready cnt:  0  latency:  0     2: Ret
#   ready list: 20
#    select 20: MachConstantBaseNode, latency:0 20  MachConstantBaseNode  ===  11  [[ 19 ]] 
#   ready list: 19
#    select 19: loadConD, latency:0 19  loadConD  === _  20  [[ 4 ]] dblcon:1.000000
#   ready list: 4
#    select 4: CallLeafDirect, latency:0 4  CallLeafDirect  ===  31  0  17  18  0  19  0  [[ 5  3 ]] dtan # double/half ( double, half ) !jvms: SDD0::foo @ bci:23 (line 10)
#   ready list: 23 25
select node_size # B4, 11
#   31: Region              31    6
#   20: MachConstantBaseNode   11 NULL NULL NULL
#   19: loadConD          NULL   20 NULL NULL
#    4: CallLeafDirect      31    0   17   18    0   19    0 NULL
#    5: MachProj             4
#    3: MachProj             4
#   35: MachProj             4
#    5: MachProj             4
#   19: loadConD          NULL   20 NULL NULL
#   20: MachConstantBaseNode   11 NULL NULL NULL
#    2: Ret                  3   21   22   18   26 NULL NULL NULL
#
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f14bf3ccf81, pid=31936, tid=31949
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8272570](https://bugs.openjdk.java.net/browse/JDK-8272570): C2: crash in PhaseCFG::global_code_motion

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Rickard Bäckman](https://openjdk.java.net/census#rbackman) (@rickard - **Reviewer**) ⚠️ Review applies to 57a88f5871fc1098c596230826bbf4cdff355445

### Contributors
 * Hui Shi `<hshi@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5140/head:pull/5140` \
`$ git checkout pull/5140`

Update a local copy of the PR: \
`$ git checkout pull/5140` \
`$ git pull https://git.openjdk.java.net/jdk pull/5140/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5140`

View PR using the GUI difftool: \
`$ git pr show -t 5140`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5140.diff">https://git.openjdk.java.net/jdk/pull/5140.diff</a>

</details>

### CODE REVIEW

**Comment by dean-long** [comment]
@casparcwang @hshi same question as the other bug: does the test assert in a debug build?

File: test/hotspot/jtreg/compiler/c2/TestGCMRecalcPressureNodes.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ */
+
+/**
+ * @test
+ * @bug 8272570
+ * @summary crash in PhaseCFG::global_code_motion
+ * @requires vm.compiler2.enabled
+ *
+ * @run main/othervm TestGCMRecalcPressureNodes
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/TestGCMRecalcPressureNodes.java:31]: Test can be executed in agent VM mode.

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
So this code assumes that the number of nodes per block does not change. Where did we add a new node?

**Comment by casparcwang** [comment]
> @casparcwang @hshi same question as the other bug: does the test assert in a debug build?

The test assert in debug build. I have upload the fast debug crash log in JBS.

**Comment by casparcwang** [comment]
> So this code assumes that the number of nodes per block does not change. Where did we add a new node?

`schedule_local` will call `sched_call`, and it will create `MachProjNode`.  That's where node `35: MachProj` comes from. 

```
#0  Node::Init (this=0x7fff5805c910, req=1) at /data/openjdk/jdk_dev/src/hotspot/share/opto/node.cpp:309
#1  0x00007ffff62206a9 in Node::Node (this=0x7fff5805c910, n0=0x7fff58059ef0) at /data/openjdk/jdk_dev/src/hotspot/share/opto/node.cpp:356
#2  0x00007ffff56c7703 in ProjNode::ProjNode (this=0x7fff5805c910, src=0x7fff58059ef0, con=8, io_use=false) at /data/openjdk/jdk_dev/src/hotspot/share/opto/multnode.hpp:68
#3  0x00007ffff56c77b6 in MachProjNode::MachProjNode (this=0x7fff5805c910, multi=0x7fff58059ef0, con=8, out=..., ideal_reg=999) at /data/openjdk/jdk_dev/src/hotspot/share/opto/machnode.hpp:739
#4  0x00007ffff605e003 in PhaseCFG::sched_call (this=0x7fff981242d0, block=0x7fff5805bc68, node_cnt=6, worklist=..., ready_cnt=..., mcall=0x7fff58059ef0, next_call=...) at /data/openjdk/jdk_dev/src/hotspot/share/opto/lcm.cpp:863
#5  0x00007ffff605f258 in PhaseCFG::schedule_local (this=0x7fff981242d0, block=0x7fff5805bc68, ready_cnt=..., next_call=..., recalc_pressure_nodes=0x0) at /data/openjdk/jdk_dev/src/hotspot/share/opto/lcm.cpp:1148
#6  0x00007ffff5cf7951 in PhaseCFG::global_code_motion (this=0x7fff981242d0) at /data/openjdk/jdk_dev/src/hotspot/share/opto/gcm.cpp:1601
#7  0x00007ffff5cf7c0c in PhaseCFG::do_global_code_motion (this=0x7fff981242d0) at /data/openjdk/jdk_dev/src/hotspot/share/opto/gcm.cpp:1642
#8  0x00007ffff5a83eac in Compile::Code_Gen (this=0x7fff98126cb0) at /data/openjdk/jdk_dev/src/hotspot/share/opto/compile.cpp:2721
#9  0x00007ffff5a7b0ac in Compile::Compile (this=0x7fff98126cb0, ci_env=0x7fff98127a20, target=0x7fff580bc9f8, osr_bci=-1, subsume_loads=true, do_escape_analysis=true, eliminate_boxing=true, do_locks_coarsening=true, install_code=true, 
    directive=0x7ffff01f1b80) at /data/openjdk/jdk_dev/src/hotspot/share/opto/compile.cpp:829
#10 0x00007ffff5966b80 in C2Compiler::compile_method (this=0x7ffff02f0910, env=0x7fff98127a20, target=0x7fff580bc9f8, entry_bci=-1, install_code=true, directive=0x7ffff01f1b80) at /data/openjdk/jdk_dev/src/hotspot/share/opto/c2compiler.cpp:107
#11 0x00007ffff5a9831a in CompileBroker::invoke_compiler_on_method (task=0x7ffff04d7c00) at /data/openjdk/jdk_dev/src/hotspot/share/compiler/compileBroker.cpp:2290
#12 0x00007ffff5a96ead in CompileBroker::compiler_thread_loop () at /data/openjdk/jdk_dev/src/hotspot/share/compiler/compileBroker.cpp:1965
#13 0x00007ffff5ab77d7 in CompilerThread::thread_entry (thread=0x7ffff02f0f30, __the_thread__=0x7ffff02f0f30) at /data/openjdk/jdk_dev/src/hotspot/share/compiler/compilerThread.cpp:59
#14 0x00007ffff6530f97 in JavaThread::thread_main_inner (this=0x7ffff02f0f30) at /data/openjdk/jdk_dev/src/hotspot/share/runtime/thread.cpp:1269
#15 0x00007ffff6530e2d in JavaThread::run (this=0x7ffff02f0f30) at /data/openjdk/jdk_dev/src/hotspot/share/runtime/thread.cpp:1252
#16 0x00007ffff652e6d2 in Thread::call_run (this=0x7ffff02f0f30) at /data/openjdk/jdk_dev/src/hotspot/share/runtime/thread.cpp:360
#17 0x00007ffff6275e17 in thread_native_entry (thread=0x7ffff02f0f30) at /data/openjdk/jdk_dev/src/hotspot/os/linux/os_linux.cpp:720
#18 0x00007ffff79b0e25 in start_thread (arg=0x7fff98128700) at pthread_create.c:308
#19 0x00007ffff74d935d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:113
```

File: test/hotspot/jtreg/compiler/c2/TestGCMRecalcPressureNodes.java (hunk 1 above)
Comment (casparcwang) [reply test/hotspot/jtreg/compiler/c2/TestGCMRecalcPressureNodes.java:31]: no, the test directly crashes with command `java TestGCMRecalcPressureNodes`. I will remove `othervm`

File: test/hotspot/jtreg/compiler/c2/TestGCMRecalcPressureNodes.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ */
+
+/**
+ * @test
+ * @bug 8272570
+ * @summary crash in PhaseCFG::global_code_motion
+ * @requires vm.compiler2.enabled
+ *
+ * @run main TestGCMRecalcPressureNodes
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/TestGCMRecalcPressureNodes.java:31]: Please move the test to `test/hotspot/jtreg/compiler/regalloc/` and the corresponding package.

File: test/hotspot/jtreg/compiler/c2/TestGCMRecalcPressureNodes.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        for (int i = 0; i < 50; i++) {
+            sh12 *= 34;
+        }
+        Math.tan(1.0);
+        bo0 = true;
+        bo1 = true;
+    }
+    public static void main(String[] args) {
+        TestGCMRecalcPressureNodes instance = new TestGCMRecalcPressureNodes();
+        for (int i = 0; i < 780000; i++) {
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/TestGCMRecalcPressureNodes.java:48]: On second thought, I think it's better to run the test with `-Xbatch` and reduce the number of iterations to `50_000` or something. That triggers the failure reliably for me.

**Review by TobiHartmann — APPROVED** [review]
Thanks for the explanation! The fix looks reasonable to me.

File: src/hotspot/share/opto/lcm.cpp [hunk 4]
```diff
@@ -518,7 +518,8 @@ Node* PhaseCFG::select(
   uint score   = 0; // Bigger is better
   int idx = -1;     // Index in worklist
   int cand_cnt = 0; // Candidate count
-  bool block_size_threshold_ok = (block->number_of_nodes() > 10) ? true : false;
+  bool block_size_threshold_ok = recalc_pressure_nodes != NULL &&
+                                 (block->number_of_nodes() > 10) ? true : false;
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/lcm.cpp:522]: And this really should be:
` block_size_threshold_ok = (recalc_pressure_nodes != NULL) && (block->number_of_nodes() > 10);`

File: test/hotspot/jtreg/compiler/c2/TestGCMRecalcPressureNodes.java (hunk 2 above)
Comment (casparcwang) [reply test/hotspot/jtreg/compiler/c2/TestGCMRecalcPressureNodes.java]: I have move it to regalloc directory.

**Comment by casparcwang** [comment]
> Thanks for the explanation! The fix looks reasonable to me.

Thank you for the review and suggestion, I have modify the code according to your suggestion.

**Review by TobiHartmann — APPROVED** [review]
Looks good.

**Review by rickard — APPROVED** [review]

**Comment by rickard** [comment]
Looks good to me too.

**Review by TobiHartmann — APPROVED** [review]

**Comment by casparcwang** [comment]
> Looks good to me too.

Thank you for your review.

**Pull request closed without merging**